spec.md

```diff
@@ -286,6 +286,67 @@ circuit Foo :
   public module Foo enablelayer A :
 ```
 
+## Formal Tests
+
+Formal tests are named top-level constructs which allow for formal test harnesses to be definied 
+as bounded model checking (BMC) problems. The body of the formal test harness is converted 
+into a BMC formula which encodes a state-transition system version of the design using one 
+of the formal backends (targetting either BTOR2 or SMTLib). 
+Formal tests have 3 operands: 
+- A symbol that the test can be referred to with.  
+- The name of the formal test harness module.  
+- A bound for the bounded model checking problem, which refers to the maximum bound k used
+to check the bounded model checking formula. This means that given a bound k, the BMC formula 
+will be checked for all unrollings of the resulting state-transition system of depths 1 through k.
+This bound is a strictly positive integer.  
+
+```firrtl
+FIRRTL version 4.0.0
+circuit Foo :
+  public module Foo :
+    ; ...
+  formal testFoo of Foo, bound = 20
+```
+More details about how bounded model checking works and what the bound refers to can be found in 
+[Biere et al. 2003](https://www.cs.cmu.edu/~emc/papers/Books%20and%20Edited%20Volumes/Bounded%20Model%20Checking.pdf).
+ 
+### Formal Test Harness  
+
+A formal test harness is a public module that is typically only used for defining the 
```

Comment:
I find the wording a little confusing. It is phrased as if we are defining a new notion ("formal test harness"), but reading on, I'm pretty sure you are just associating a module definition with the `formal` block.

Reply:
I just want to describe what the expected input is. Rather than giving the formal block a region, we instead reuse the notion of a module to just wrap whatever logic the test will contain. In the module given to a formal test, inputs are considered symbolic/free, which allows us to avoid introducing a new construct call `symbolic` that does exactly that. This is just conceptual reuse but I wanted to describe how the body of the module given to a formal op differs from a regular module body.

Reply:
Could I recommend this simpler variation?:

> A public module definition may be used as a formal test harness. Their `input` ports act as symbolic variables for the test.

Again, this sticks to just the mechanics of the feature, and avoids commenting on how it should be used. (If we want to include suggestions like "a test harness should instantiate a DUT" etc, we should split it out into its own paragraph).


By the way, does this mean the `symbolic` statements have been removed?

Reply:
Yes @darthscsi pointed out that if these behave like undriven input then they should look like undriven inputs and that I should not add a new op that simply reproduces the behavior of another op but in a different context.